crates/pallet-currency-swap/src/lib.rs

```diff
@@ -0,0 +1,140 @@
+    <T as frame_system::Config>::AccountId,
+    <T as Config>::AccountIdTo,
+>>::From;
+
+/// Utility alias for easy access to the [`Currency::Balance`] of
+/// the [`primitives_currency_swap::CurrencySwap::From`] type.
+type FromBalanceOf<T> =
+    <FromCurrencyOf<T> as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// Utility alias for easy access to [`primitives_currency_swap::CurrencySwap::To`] type from a given config.
+type ToCurrencyOf<T> = <<T as Config>::CurrencySwap as CurrencySwapT<
+    <T as frame_system::Config>::AccountId,
+    <T as Config>::AccountIdTo,
+>>::To;
+
+/// Utility alias for easy access to the [`Currency::Balance`] of
+/// the [`primitives_currency_swap::CurrencySwap::To`] type.
+type ToBalanceOf<T> = <ToCurrencyOf<T> as Currency<<T as Config>::AccountIdTo>>::Balance;
+
+// We have to temporarily allow some clippy lints. Later on we'll send patches to substrate to
+// fix them at their end.
+#[allow(clippy::missing_docs_in_private_items)]
+#[frame_support::pallet]
+pub mod pallet {
+    use frame_support::{
+        pallet_prelude::*,
+        storage::with_storage_layer,
+        traits::{ExistenceRequirement, Imbalance, WithdrawReasons},
+    };
+    use frame_system::pallet_prelude::*;
+    use sp_runtime::traits::MaybeDisplay;
+    use sp_std::fmt::Debug;
+
+    use super::*;
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    /// Configuration trait of this pallet.
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// The user account identifier type to convert to.
+        type AccountIdTo: Parameter
+            + Member
+            + MaybeSerializeDeserialize
+            + Debug
+            + MaybeDisplay
+            + Ord
+            + MaxEncodedLen;
+
+        /// Interface into currency swap implementation.
+        type CurrencySwap: CurrencySwapT<Self::AccountId, Self::AccountIdTo>;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Balances were swapped.
+        BalancesSwapped {
+            /// The account id balances withdrawed from.
+            from: T::AccountId,
+            /// The withdrawed balances amount.
+            withdrawed_amount: FromBalanceOf<T>,
+            /// The account id balances deposited to.
+            to: T::AccountIdTo,
+            /// The deposited balances amount.
+            deposited_amount: ToBalanceOf<T>,
+        },
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Swap balances.
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::swap())]
+        pub fn swap(
```

Comment:
We might want to add a `swap_keep_alive` call here, to expose a `ExistenceRequirement::KeepAlive` variant of this call.

Reply:
https://github.com/humanode-network/humanode/issues/689